test/cpp/end2end/client_lb_end2end_test.cc

```diff
@@ -2882,43 +2890,55 @@ TEST_F(ControlPlaneStatusRewritingTest, RewritesFromConfigSelector) {
 // WeightedRoundRobinTest
 //
 
-using WeightedRoundRobinTest = ClientLbEnd2endTest;
-
-TEST_F(WeightedRoundRobinTest, Basic) {
+const char kServiceConfigPerCall[] =
+    "{\n"
+    "  \"loadBalancingConfig\": [\n"
+    "    {\"weighted_round_robin_experimental\": {\n"
+    "      \"blackoutPeriod\": \"0s\"\n"
+    "    }}\n"
+    "  ]\n"
+    "}";
+
+const char kServiceConfigOob[] =
+    "{\n"
+    "  \"loadBalancingConfig\": [\n"
+    "    {\"weighted_round_robin_experimental\": {\n"
+    "      \"blackoutPeriod\": \"0s\",\n"
+    "      \"enableOobLoadReport\": true\n"
+    "    }}\n"
+    "  ]\n"
+    "}";
+
+class WeightedRoundRobinParamTest
+    : public ClientLbEnd2endTest,
+      public ::testing::WithParamInterface<const char*> {};
+
+TEST_P(WeightedRoundRobinParamTest, Basic) {
   const int kNumServers = 3;
   StartServers(kNumServers);
-  // Tell each server to report the appropriate CPU utilization.
-  xds::data::orca::v3::OrcaLoadReport load_report;
-  load_report.set_rps_fractional(100);
-  load_report.set_cpu_utilization(0.9);
-  servers_[0]->service_.SetLoadReport(load_report);
-  load_report.set_cpu_utilization(0.3);
-  servers_[1]->service_.SetLoadReport(load_report);
-  servers_[2]->service_.SetLoadReport(load_report);
+  // Report server metrics that should give 1:3:3 WRR picks.
+  servers_[0]->server_metric_recorder_.SetCpuUtilization(0.9);
+  servers_[0]->server_metric_recorder_.SetQps(100);
+  servers_[1]->server_metric_recorder_.SetCpuUtilization(0.3);
+  servers_[1]->server_metric_recorder_.SetQps(100);
+  servers_[2]->server_metric_recorder_.SetCpuUtilization(0.3);
+  servers_[2]->server_metric_recorder_.SetQps(100);
   // Create channel.
   auto response_generator = BuildResolverResponseGenerator();
   auto channel = BuildChannel("", response_generator);
   auto stub = BuildStub(channel);
-  const char kServiceConfig[] =
-      "{\n"
-      "  \"loadBalancingConfig\": [\n"
-      "    {\"weighted_round_robin_experimental\": {\n"
-      "      \"blackoutPeriod\": \"0s\"\n"
-      "    }}\n"
-      "  ]\n"
-      "}";
-  response_generator.SetNextResolution(GetServersPorts(), kServiceConfig);
+  response_generator.SetNextResolution(GetServersPorts(), GetParam());
   // Wait for the right set of WRR picks.
   size_t num_picks = 0;
   SendRpcsUntil(DEBUG_LOCATION, stub, [&](const Status&) {
-    if (++num_picks == 7) {
+    if (++num_picks == 21) {
```

Comment:
I think we want to be more rigorous here than just checking that the aggregate counts are right after 3 full passes -- we actually need to check that each individual pass has the right counts.  If the policy wound up effectively combining all three passes into one, the aggregate counts would be fine, but the data in one individual pass would not be okay, and the test would not detect that.

I suggest something like this:

```
[&](const Status&) {
  if (++num_picks == 7) {
    if (servers_[0]->service_.request_count() == 1 &&
        servers_[1]->service_.request_count() == 3 &&
        servers_[2]->service_.request_count() == 3) {
      if (++num_passes == 3) return false;
    } else {
      num_passes = 0;
    }
    num_picks = 0;
    ResetCounters();
  }
  return true;
}
```

Note that this lambda is now duplicated in about 3 different places, so it might make sense to refactor it into a helper function, parameterized by the expected counts to see for each server in each pass.

Reply:
Updated. We now validate 3 consecutive passes but with sleep(weight_update_period) in between, otherwise it blasts ~100 validations before a weight update.

Regarding refactoring, I feel it's too dry (go/tott/598) for tests. The lambda has captured vars so I could only think of an option to refactor SendRpcUntil to accept a stateful functor. Also I expect we don't want to hardcode with 3 servers, then it already becomes a bit too hard to read for tests.

Reply:
I think that TotT is primarily talking about making sure that all of the data that affects a test is defined in the test itself.  But here we're not talking about data, we're talking about code.  I think it's totally reasonable to have helper functions for common operations like this, especially if they're represented in an API that makes the test more readable.

I think you can do this as follows:

```
class WeightedRoundRobinTest : public ClientLbEnd2endTest {
 protected:
  void ExpectWeightedRoundRobinPicks(
      const grpc_core::DebugLocation& location,
      const std::unique_ptr<grpc::testing::EchoTestService::Stub>& stub,
      const std::vector<size_t>& expected_weights, size_t total_passes) {
    GPR_ASSERT(expected_weights.size() == servers_.size());
    size_t total_picks_per_pass = 0;
    for (size_t picks : expected_weights) {
      total_picks_per_pass += picks;
    }
    size_t num_picks = 0;
    size_t num_passes = 0;
    SendRpcsUntil(
        location, stub, [&](const Status&) {
          if (++num_picks == total_picks_per_pass) {
            bool match = true;
            for (size_t i = 0; i < expected_weights.size(); ++i) {
              if (servers_[i]->service_.request_count() != expected_weights[i]) {
                match = false;
                break;
              }
            }
            if (match) {
              if (++num_passes = total_passes) return false;
            } else {
              num_passes = 0;
            }
            num_picks = 0;
            ResetCounters();
          }
          return true;
        });
  }
};

class WeightedRoundRobinParamTest
    : public WeightedRoundRobinTest,
      public ::testing::WithParamInterface<const char*> {};
```

Then, in the individual tests, you can do something like this:

```
ExpectWeightedRoundRobinPicks(
    DEBUG_LOCATION, stub, {1, 3, 3}, 3);
```

Reply:
Done.